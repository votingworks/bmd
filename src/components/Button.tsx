import React, { MouseEventHandler } from 'react'
import styled, { css } from 'styled-components'

interface Attrs extends HTMLButtonElement {
  readonly type: string
}

export interface ButtonInterface {
  readonly danger?: boolean
  readonly primary?: boolean
  readonly fullWidth?: boolean
  readonly small?: boolean
  readonly big?: boolean
}

interface Props
  extends ButtonInterface,
    React.PropsWithoutRef<JSX.IntrinsicElements['button']> {}

const buttonStyles = css<Props>`
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: ${({ danger = false, primary = false }) =>
    (danger && 'red') ||
    (primary && 'rgb(71, 167, 75)') ||
    'rgb(211, 211, 211)'};
  cursor: ${({ disabled = false }) => (disabled ? undefined : 'pointer')};
  width: ${({ fullWidth = false }) => (fullWidth ? '100%' : undefined)};
  padding: ${({ big = false, small = false }) =>
    // eslint-disable-next-line no-nested-ternary
    small ? '0.35rem 0.5rem' : big ? '1rem 1.75rem' : '0.75rem 1rem'};
  line-height: 1.25;
  color: ${({ disabled = false, danger = false, primary = false }) =>
    (disabled && 'rgb(169, 169, 169)') ||
    (danger && '#FFFFFF') ||
    (primary && '#FFFFFF') ||
    'black'};
  font-size: ${({ big = false }) => (big ? '1.25rem' : undefined)};
  touch-action: manipulation;
`

export const DecoyButton = styled.div`
  ${buttonStyles} /* stylelint-disable-line value-keyword-case */
`

const StyledButton = styled('button').attrs((props: Attrs) => ({
  type: props.type || 'button',
}))`
  ${buttonStyles} /* stylelint-disable-line value-keyword-case */
`

interface PointerButtonProps extends Props {
  onPress: MouseEventHandler
}

const Button = ({ onPress, ...rest }: PointerButtonProps) => (
  <StyledButton {...rest} onClick={onPress} />
)

export const SegmentedButton = styled.span`
  display: inline-block;
  white-space: nowrap;
  & > button:not(:last-child) {
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }
  & > button:not(:first-child) {
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }
  & > button:disabled {
    background: #028099;
    color: #ffffff;
  }
`

export default Button
