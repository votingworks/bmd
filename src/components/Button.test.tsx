import { axe } from 'jest-axe'
import React from 'react'
import { render } from 'react-testing-library'

import Button from './Button'

it(`renders Button`, async () => {
  const { container } = render(<Button>foo</Button>)
  expect(await axe(container.innerHTML)).toHaveNoViolations()
  expect(container.firstChild).toMatchSnapshot()
})

it(`renders primary Button`, async () => {
  const { container } = render(<Button primary>Primary</Button>)
  expect(await axe(container.innerHTML)).toHaveNoViolations()
  expect(container.firstChild).toMatchSnapshot()
})

it(`renders danger Button`, async () => {
  const { container } = render(<Button danger>Danger!</Button>)
  expect(await axe(container.innerHTML)).toHaveNoViolations()
  expect(container.firstChild).toMatchSnapshot()
})
