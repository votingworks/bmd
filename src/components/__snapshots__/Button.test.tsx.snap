// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Button 1`] = `
.c0 {
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: rgb(211,211,211);
  cursor: pointer;
  padding: 0.75rem 1rem;
  line-height: 1.25;
  color: black;
  touch-action: manipulation;
}

<button
  class="c0"
  type="button"
>
  foo
</button>
`;

exports[`renders DecoyButton 1`] = `
.c0 {
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: rgb(211,211,211);
  cursor: pointer;
  padding: 0.75rem 1rem;
  line-height: 1.25;
  color: black;
  touch-action: manipulation;
}

<div
  class="c0"
>
  DecoyButton
</div>
`;

exports[`renders big Button 1`] = `
.c0 {
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: rgb(211,211,211);
  cursor: pointer;
  padding: 1rem 1.75rem;
  line-height: 1.25;
  color: black;
  font-size: 1.25rem;
  touch-action: manipulation;
}

<button
  class="c0"
  type="button"
>
  I’m a big button!
</button>
`;

exports[`renders danger Button 1`] = `
.c0 {
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: red;
  cursor: pointer;
  padding: 0.75rem 1rem;
  line-height: 1.25;
  color: #FFFFFF;
  touch-action: manipulation;
}

<button
  class="c0"
  type="button"
>
  Danger!
</button>
`;

exports[`renders primary Button 1`] = `
.c0 {
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: rgb(71,167,75);
  cursor: pointer;
  padding: 0.75rem 1rem;
  line-height: 1.25;
  color: #FFFFFF;
  touch-action: manipulation;
}

<button
  class="c0"
  type="button"
>
  Primary
</button>
`;

exports[`renders small Button 1`] = `
.c0 {
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: rgb(211,211,211);
  cursor: pointer;
  padding: 0.35rem 0.5rem;
  line-height: 1.25;
  color: black;
  touch-action: manipulation;
}

<button
  class="c0"
  type="button"
>
  I’m a small button!
</button>
`;
